rust/noosphere-ns/src/name_system.rs

```diff
@@ -0,0 +1,195 @@
+use ucan_key_support::ed25519::Ed25519KeyMaterial;
+
+/// The [NameSystem] is responsible for both propagating and resolving address records
+/// in the Noosphere DHT. Hosted records can be set via `set_record(identity, link)`, propagating the
+/// record immediately, and repropagated every `propagation_interval` seconds. Records
+/// can be resolved via `get_record(identity)`.
+pub struct NameSystem<'a> {
+    /// Bootstrap peers for the DHT network.
+    pub(crate) bootstrap_peers: Option<&'a Vec<Multiaddr>>,
+    pub(crate) dht: Option<DHTNode>,
+    pub(crate) dht_config: DHTConfig,
+    /// Key of the NameSystem's sphere.
+    pub(crate) key_material: &'a Ed25519KeyMaterial,
+    /// In seconds, the Time-To-Live (TTL) duration of records set
+    /// in the network, and implicitly, how often the records are
+    /// propagated.
+    pub(crate) ttl: u64,
+    /// Map of sphere DIDs to [NSRecord] hosted/propagated by this name system.
+    pub(crate) hosted_records: HashMap<String, NSRecord>,
+    /// Map of resolved sphere DIDs to resolved [NSRecord].
+    pub(crate) resolved_records: HashMap<String, NSRecord>,
+}
+
+impl<'a> NameSystem<'a> {
+    /// Initializes and attempts to connect to the network.
+    pub async fn connect(&mut self) -> Result<()> {
+        let mut dht = DHTNode::new(self.key_material, self.bootstrap_peers, &self.dht_config)?;
+        dht.run().map_err(|e| anyhow!(e.to_string()))?;
+        dht.bootstrap().await.map_err(|e| anyhow!(e.to_string()))?;
+        dht.wait_for_peers(1)
+            .await
+            .map_err(|e| anyhow!(e.to_string()))?;
+        self.dht = Some(dht);
+        Ok(())
+    }
+
+    /// Disconnect and deallocate connections to the network.
+    pub fn disconnect(&mut self) -> Result<()> {
+        if let Some(mut dht) = self.dht.take() {
+            dht.terminate()?;
+        }
+        Ok(())
+    }
+
+    /// Propagates all hosted records on nearby peers in the DHT network.
+    /// in the DHT network. Automatically called every `ttl` seconds (TBD),
+    /// but can be manually called push updated records to the network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn propagate_records(&self) -> Result<()> {
+        if self.dht.is_none() {
+            return Err(anyhow!("not connected"));
+        }
+        if self.hosted_records.is_empty() {
+            return Ok(());
+        }
+
+        let pending_tasks: Vec<_> = self
+            .hosted_records
+            .iter()
+            .map(|(identity, record)| self.dht_set_record(identity, record))
+            .collect();
+        try_join_all(pending_tasks).await?;
+        Ok(())
+    }
+
+    /// Propagates the corresponding managed sphere's content Cid on nearby peers
+    /// in the DHT network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn set_record(&mut self, identity: &String, link: &Cid) -> Result<()> {
```

Comment:
Similarly to `get_record`, this should also probably take a `NSRecord` instead of a `Cid` as we'll be propagating 2-3 values at least per record.

Reply:
(discussed offline, no action required)